Documentation/upgrading-from-ipy2.md

```diff
@@ -14,3 +14,167 @@ In an effort to improve compatibility, `sys.platform` no longer returns `cli`. I
+
+The Python `int` type in IronPython 3 is implemented as `System.Numerics.BigInteger` (and not as `System.Int32` as it was in IronPython 2). It can contain in theory an arbitrarily large integer (only limited by the 2 GByte memory boundary).
+
+```pycon
+>>> import clr
+>>> clr.AddReference("System.Numerics")
+>>> import System
+>>> int is System.Numerics.BigInteger
+True
+>>> int is System.Int32
+False
+>>> clr.GetClrType(int).Name
+'BigInteger'
+```
+
+This means that in interop cases, when the `int` type is used (think generics), it will mean `BigInteger` and not `Int32` (which was the case in IronPython 2). For example, `System.Collections.Generic.List[int]` in IronPython 3 is equivalent to `System.Collections.Generic.List[System.Numerics.BigInteger]`. To retain IronPython 2 semantics, replace `int` with `System.Int32`.
+
+Overview of `int` type equivalency:
+
+| IronPython 2 | IronPython 3 | .NET                         |
+| ------------ | ------------ | ---------------------------- |
+| `long`       | `int`        | `System.Numerics.BigInteger` |
+| `int`        | N/A          | `System.Int32`               |
+
+### Instances of `int`
+
+As for instances of `int`, mostly for performance reasons, IronPython may use instances of `System.Int32` to hold smaller integers, while `BigInteger` instances are used for large integers. This is done transparently from the Python side, but again the distinction may become relevant for interop cases. Examples:
+
+```python
+i = 1        # instance of Int32
+j = 1 << 31  # instance of BigInteger
+k = j - 1    # still BigInteger, as one of the arguments makes the result type BigInteger
+```
+
+This means that the type of `Int32` objects is always reported as `int` (which is the same as `BigInteger`). If it is important to check what is the actual type of a given integer object, test for the presence of `MaxValue` or `MinValue`. For those properties to be visible, `System` has to be imported first.
+
+```pycon
+>>> import System
+>>> type(i)
+<class 'int'>
+>>> hasattr(i, 'MaxValue') # Int32
+True
+>>> hex(i.MaxValue)
+'0x7fffffff'
+>>> type(j)
+<class 'int'>
+>>> hasattr(j, 'MaxValue') # BigInteger
+False
+```
+
+The creation of either `Int32` or `BigInteger` instances happens automatically by the `int` constructor. If for interop purposes it is important to create a `BigInteger` (despite the value fitting in 32 bits), use method `ToBigInteger`. It converts `Int32` values to `BigInteger` and leaves `BigInteger` values unaffected.
+
+```pycon
+>>> bi = i.ToBigInteger()
+>>> hasattr(bi, 'MaxValue')
+False
+```
+
+In the opposite direction, if it is essential to create `Int32` objects, either use constructors for `int` or `Int32`. The former converts an integer to `Int32` if the value fits in 32 bits, otherwise it leaves it as `BigInteger`. The latter throws an exception is the conversion is not possible.
```

Comment:
I wonder if we should avoid suggesting the `int` constructor in case future selves decide that we want to change the behaviour (to not force to `Int32` when in range). Or at least put some sort of disclaimer (like "in the current implementation").

Reply:
Being able of "normalizing" an instance in some way is important. From the CLR perspective, `BigInteger(1)` is not the same object as `Int32(1)`. I was thinking of using `__pos__` (operator `+`) for that purpose, but somehow the constructor felt like the most appropriate way. Indeed, it may change in the future, like if `Int64` is added to the list of types we use to accelerate `int`. Or, going in the other direction, if `Int32` is dropped and `int` is just `BigInteger` all the time.

I am fine with the disclaimer "in the current implementation".